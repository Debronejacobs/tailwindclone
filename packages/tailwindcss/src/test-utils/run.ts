import { optimizeCss } from '@tailwindcss/optimize'
import { compile } from '..'

export function compileCss(css: string, candidates: string[] = []) {
  return optimizeCss(compile(css).build(candidates)).trim()
}

export function run(candidates: string[]) {
  return optimizeCss(compile('@tailwind utilities;').build(candidates)).trim()
}
