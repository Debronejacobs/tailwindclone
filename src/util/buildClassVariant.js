import escapeClassName from './escapeClassName'

export default function buildClassVariant(className, variantName, variantsAsSuffix, separator) {
  if (variantsAsSuffix) {
    return `${className}${escapeClassName(separator)}${variantName}`
  }
  return `.${variantName}${escapeClassName(separator)}${className.slice(1)}`
}
