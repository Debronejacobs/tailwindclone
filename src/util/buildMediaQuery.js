import _ from 'lodash'

function isInvalidValue(value) {
  return !value || value === '' || value === 0 || value === '0'
}

export default function buildMediaQuery(screens) {
  if (isInvalidValue(screens)) {
    return undefined
  }

  if (_.isString(screens)) {
    screens = { min: screens }
  }

  if (!Array.isArray(screens)) {
    screens = [screens]
  }

  return _(screens)
    .map((screen) => {
      if (_.has(screen, 'raw')) {
        return screen.raw
      }

      return _(screen)
        .map((value, feature) => {
          feature = _.get(
            {
              min: 'min-width',
              max: 'max-width',
            },
            feature,
            feature
          )

          if (isInvalidValue(value)) {
            return undefined
          }

          return `(${feature}: ${value})`
        })
        .uniq()
        .filter((v) => v !== undefined)
        .join(' and ')
    })
    .join(', ')
}
