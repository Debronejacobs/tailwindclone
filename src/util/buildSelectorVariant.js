import parser from 'postcss-selector-parser'
import tap from 'lodash/tap'

export default (selector, variantName, separator, onError = () => {}) =>
  parser(selectors => {
    tap(selectors.first.filter(({ type }) => type === 'class').pop(), classSelector => {
      if (classSelector === undefined) {
        onError('Variant cannot be generated because selector contains no classes.')
        return
      }

      classSelector.value = `${variantName}${separator}${classSelector.value}`
    })
  }).processSync(selector)
