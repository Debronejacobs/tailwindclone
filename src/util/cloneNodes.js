import _ from 'lodash'

export default nodes => _.map(nodes, node => node.clone())
