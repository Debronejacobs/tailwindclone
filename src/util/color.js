import namedColors from 'color-name'

let HEX = /^#([a-f\d]{2})([a-f\d]{2})([a-f\d]{2})([a-f\d]{2})?$/i
let SHORT_HEX = /^#([a-f\d])([a-f\d])([a-f\d])([a-f\d])?$/i
let VALUE = `(?:\\d+|\\d*\\.\\d+)%?`
let SEP = `(?:,\\s*|\\s+)`
let ALPHA_SEP = `\\s*[,/]\\s*`
let RGB_HSL = new RegExp(
  `^(rgb|hsl)a?\\(\\s*(${VALUE})${SEP}(${VALUE})${SEP}(${VALUE})(?:${ALPHA_SEP}(${VALUE}))?\\s*\\)$`
)

export function parseColor(value) {
  if (typeof value !== 'string') {
    return null
  }

  value = value.trim()

  if (value in namedColors) {
    return {
      mode: 'rgb',
      color: namedColors[value],
    }
  }

  let hex = value
    .replace(SHORT_HEX, (_, r, g, b, a) => ['#', r, r, g, g, b, b, a ? a + a : ''].join(''))
    .match(HEX)

  if (hex !== null) {
    return {
      mode: 'rgb',
      color: [parseInt(hex[1], 16), parseInt(hex[2], 16), parseInt(hex[3], 16)],
      alpha: hex[4] ? parseInt(hex[4], 16) / 255 : undefined,
    }
  }

  let match = value.match(RGB_HSL)

  if (match !== null) {
    return {
      mode: match[1],
      color: [match[2], match[3], match[4]],
      alpha: match[5],
    }
  }

  return null
}

export function formatColor({ mode, color, alpha }) {
  let hasAlpha = alpha !== undefined
  return `${mode}${hasAlpha ? 'a' : ''}(${color.join(', ')}${hasAlpha ? `, ${alpha}` : ''})`
}
