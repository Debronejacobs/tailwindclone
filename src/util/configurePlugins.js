export default (pluginConfig, plugins) => {
  const pluginNames = Array.isArray(pluginConfig)
    ? pluginConfig
    : Object.keys(plugins).filter(
        pluginName => pluginConfig !== false && pluginConfig[pluginName] !== false
      )

  return pluginNames.map(pluginName => plugins[pluginName]())
}
