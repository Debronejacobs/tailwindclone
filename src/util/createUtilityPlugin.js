import _ from 'lodash'
import fromPairs from 'lodash/fromPairs'
import toPairs from 'lodash/toPairs'
import castArray from 'lodash/castArray'

function className(classPrefix, key) {
  if (key === '') {
    return classPrefix
  }

  if (key.startsWith('-')) {
    return `-${classPrefix}${key}`
  }

  return `${classPrefix}-${key}`
}

export default function createUtilityPlugin(themeKey, utilityVariations, mapTheme = theme => theme) {
  return function({ e, addUtilities, variants, theme }) {
    const utilities = utilityVariations.map(([classPrefix, properties]) => {
      return fromPairs(
        toPairs(mapTheme(theme(themeKey))).map(([key, value]) => {
          return [
            `.${e(className(classPrefix, key))}`,
            fromPairs(castArray(properties).map(property => [property, value])),
          ]
        })
      )
    })
    return addUtilities(utilities, variants(themeKey))
  }
}

export function convertDefaultKey(theme) {
  return _.mapKeys(theme, (value, key) => key === 'default' ? '' : key)
}
