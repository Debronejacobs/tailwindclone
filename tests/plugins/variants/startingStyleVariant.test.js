import { quickVariantPluginTest } from '../../util/run'

quickVariantPluginTest('startingStyleVariant').toMatchSnapshot()
