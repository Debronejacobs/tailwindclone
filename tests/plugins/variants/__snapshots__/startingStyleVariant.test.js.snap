// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should test the 'startingStyleVariant' plugin 1`] = `
"
@starting-style {
  .starting\\:flex {
    display: flex;
  }
}
"
`;
